File: workspace/RainBackend/src/com/lucky/watisrain/backend/data/Map.java
package com.lucky.watisrain.backend.data;

import java.util.ArrayList;
import java.util.List;

import com.lucky.watisrain.backend.Util;

/**
 * A Map contains a set of Locations and Paths. It represents the graph data.
 */
public class Map {

	private List<Building> buildings;
	
	// all locations, active and passive
	private List<Location> locations;
	
	private List<Path> paths;
	
	/**
	 * Initialize empty Map object
	 */
	public Map(){
		buildings = new ArrayList<>();
		locations = new ArrayList<>();
		paths = new ArrayList<>();
	}
	
	/*
	 * Add all floors of a building, and all its stairs
	 */
	public void addBuilding(Building building){
		buildings.add(building);
		locations.addAll(building.getAllFloors());
		paths.addAll(building.getAllStairs());
	}
	
	public void addPath(Path path){
		paths.add(path);
	}
	
	public void addPassiveLocation(Location loc){
		locations.add(loc);
	}
	
	/**
	 * Return all of the paths in the graph
	 */
	public List<Path> getPaths(){
		return paths;
	}
	
	/**
	 * Fetch the path object between two Locations, if exists.
	 * Otherwise, return null
	 */
	public Path retrievePath(Location a, Location b){
		for(Path path : paths){
			if(path.getPointA().equals(a) && path.getPointB().equals(b))
				return path;
			if(path.getPointA().equals(b) && path.getPointB().equals(a))
				return path;
		}
		return null;
	}
	
	public List<Location> getLocations(){
		return locations;
	}
	
	public List<Building> getBuildings(){
		return buildings;
	}
	
	
	/**
	 * Given the unique identifier of a building, return the Building object
	 */
	public Building getBuildingByID(String name){
		for(Building building : buildings){
			if(building.getName().equals(name))
				return building;
		}
		return null;
	}
	
	
	/**
	 * Try to fetch some form of location from a string by trying a variety
	 * of methods
	 */
	public Location getLocationByID(String name){
		
		// Attempt to parse as building
		Building building = getBuildingByID(name);
		if(building != null)
			return building.getMainFloor();
		
		// Attempt to parse as-is (usually combined ID)
		for(Location loc : locations){
			if(loc.getName().equals(name))
				return loc;
		}
		
		// Final attempt: parse only the building half
		String partBuilding = Util.getBuilding(name);
		for(Location loc : locations){
			if(loc.getName().equals(partBuilding))
				return loc;
		}
		
		return null;
	}
	
	
	/**
	 * Debugging purposes
	 */
	public void printDataToStdout(){
		
		System.out.println("Buildings:");
		for(Building b : buildings){
			System.out.println(b);
		}
		System.out.println();
		
		System.out.println("Locations:");
		for(Location l : locations){
			System.out.println(l);
		}
		System.out.println();
		
		System.out.println("Paths:");
		for(Path p : paths){
			System.out.println(p);
		}
		System.out.println();
		
	}
	
}

File: iOS/WatBackend/WatBackend/Map.m

#import <Foundation/Foundation.h>
#import "Map.h"
#import "Util.h"

@implementation Map

- (Map*) init{
    _buildings = [[NSMutableArray alloc] init];
    _locations = [[NSMutableArray alloc] init];
    _paths = [[NSMutableArray alloc] init];
    return self;
}

- (void) addBuilding:(Building *)building{
    [self.buildings addObject:building];
    [self.locations addObjectsFromArray:[building getAllFloors]];
    [self.paths addObjectsFromArray:[building getAllStairs]];
}

- (void) addPassiveLocation:(id)location{
    [self.locations addObject:location];
}

- (void) addPath:(Path *)path{
    [self.paths addObject:path];
}

- (Building*) getBuildingByID:(NSString*) name{
    for(Building *building in _buildings){
        if([[building name] isEqual:name])
            return building;
    }
    return nil;
}

- (Location*) getLocationByID:(NSString*) name{
    Building *building = [self getBuildingByID:name];
    if(building != nil)
        return [building getMainFloor];
    
    for(Location *loc in _locations){
        if([[loc name] isEqual:name])
            return loc;
    }
    
    NSString *partBuilding = getBuilding(name);
    for(Location *loc in _locations){
        if([[loc name] isEqual:partBuilding])
            return loc;
    }
    
    return nil;
}

- (Path*) retrievePathFrom:(Location *)a To:(Location *)b{
    for(Path *path in _paths){
        if([[path pointA] isEqual:a] && [[path pointB] isEqual:b])
            return path;
        if([[path pointB] isEqual:a] && [[path pointA] isEqual:b])
            return path;
    }
    return nil;
}

@end
